### PR TITLE
Generate command updates 

### PR DESCRIPTION
**Pull request by dnitsch**
merge environments correctly with precedence

### CODE REVIEW

File: docs/v2.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# V2 migration
+
+Changes:
+
+- Context 
+    - container first class citizen 
+    - default mount point in the container is /workspace/.taskctl
+        - if you are referencing an absolute path this
```
Comment (ElvenSpellmaker) [inline docs/v2.md:8]: Missing README part, but also I think we might need to change this back until we have everything migrated across really

File: internal/config/pipeline.go [hunk 2]
```diff
@@ -80,7 +80,7 @@ func buildPipeline(g *scheduler.ExecutionGraph, stages []*PipelineDefinition, cf
 		if _, err := g.Node(stage.Name); err == nil {
 			return nil, fmt.Errorf("%w, stage with same name %s already exists", ErrStageBuildFailure, stage.Name)
 		}
-
+		// TODO: investigate gha sample p2 pipeline incorrect chilred adding
```
Comment (dnitsch) [inline internal/config/pipeline.go:83]: fixed - removed

File: cmd/taskctl/testdata/gha.sample.yml [hunk 3]
```diff
@@ -81,7 +81,8 @@ pipelines:
     - pipeline: infrapipeline
       env:
         ENV_NAME: nonprod
-        SOME_TOKEN_FOR_APP1: fuzzzbarsdsd
+        # locall you can keep this value in a taskctl.env file
```
Comment (ElvenSpellmaker) [inline cmd/taskctl/testdata/gha.sample.yml:84]:
```suggestion
        # locally you can keep this value in a taskctl.env file
```

File: internal/config/task.go [hunk 4]
```diff
@@ -24,6 +25,16 @@ func buildTask(def *TaskDefinition, lc *loaderContext) (*task.Task, error) {
...
 	t.Env = variables.FromMap(def.Env).Merge(t.Env)
+	t.EnvFile = utils.NewEnvFile(func(e *utils.Envfile) {
+		if def.Envfile != nil {
+			e.Exclude = def.Envfile.Exclude
+			e.Include = def.Envfile.Include
+			e.Modify = def.Envfile.Modify
+			e.Path = def.Envfile.Path
+			e.Quote = def.Envfile.Quote
+			e.ReplaceChar = def.Envfile.ReplaceChar
+		}
```
Comment (ElvenSpellmaker) [inline internal/config/task.go:36]: Is that really the right way to do this, is there no clone or similar?

File: internal/genci/githubimpl.go [hunk 5]
```diff
@@ -133,13 +135,23 @@ func convertTaskToStep(node *scheduler.Stage) *schema.GithubStep {
 		if gh.If != "" {
 			step.If = gh.If
 		}
+		// if env is specified on this level we want to overwrite it
+		// with ci_meta.github.env keys
+		if gh.Env != nil {
+			step.Env = node.Env().
+				Merge(node.Task.Env).
+				Merge(variables.FromMap(utils.ConvertToMapOfStrings(gh.Env))).
+				Map()
```
Comment (ElvenSpellmaker) [inline internal/genci/githubimpl.go:144]: What does this do without any args?

File: pkg/runner/runner.go [hunk 6]
```diff
@@ -148,6 +148,18 @@ func (r *TaskRunner) Run(t *task.Task) error {
 	env := r.env.Merge(execContext.Env)
 	env = env.With("TASK_NAME", t.Name)
 	env = env.Merge(t.Env)
+	// denormalized graph will append all ancestral env keys to the task
+	// if task also includes an envfile property
+	// We need to read it in and hang on the env for the command compiler.
+	if reader, exists := utils.ReaderFromPath(t.EnvFile); exists {
+		m, err := utils.ReadEnvFile(reader)
+		if err != nil {
+			// TODO: specify error wrapping here
```
Comment (ElvenSpellmaker) [inline pkg/runner/runner.go:157]: Yeah?

File: pkg/scheduler/graph.go [hunk 7]
```diff
@@ -172,7 +184,7 @@ func (g *ExecutionGraph) BFSNodesFlattened(nodeName string) []*Stage {
 		current := queue[0]
 		queue = queue[1:]
 
-		// add to flattened list - except if it's the root node
+		// // add to flattened list - except if it's the root node
```
Comment (ElvenSpellmaker) [inline pkg/scheduler/graph.go:187]:
```suggestion
		// add to flattened list - except if it's the root node
```

File: pkg/scheduler/stage.go [hunk 8]
```diff
@@ -81,6 +81,17 @@ func (s *Stage) FromStage(originalStage *Stage, existingGraph *ExecutionGraph, a
...
+		tsk.EnvFile = utils.NewEnvFile(func(e *utils.Envfile) {
+			if tsk.EnvFile != nil {
+				e.Exclude = tsk.EnvFile.Exclude
+				e.Include = tsk.EnvFile.Include
+				e.Modify = tsk.EnvFile.Modify
+				e.Path = tsk.EnvFile.Path
+				e.Quote = tsk.EnvFile.Quote
+				e.ReplaceChar = tsk.EnvFile.ReplaceChar
+			}
+		})
```
Comment (ElvenSpellmaker) [inline pkg/scheduler/stage.go:94]: Likewise are we sure this is the right thing here?

File: tasks.yaml [hunk 9]
```diff
@@ -198,3 +198,4 @@ tasks:
     name: "tidy me up"
     command:
       - go mod tidy
+      - echo $SET_IN_TASKCTL_DEFAULT
```
Comment (ElvenSpellmaker) [inline tasks.yaml:201]: huh?

File: internal/config/task.go (hunk 4 above)
Comment (dnitsch) [reply internal/config/task.go]: I suppose we already have mergo as a dependency - so we could use that...

File: internal/genci/githubimpl.go (hunk 5 above)
Comment (ElvenSpellmaker) [reply internal/genci/githubimpl.go:144]: Oh, I now see the `.` on the previous line...

Comment (dnitsch) [reply internal/genci/githubimpl.go:144]: ah yes sorry - wasn't sure what you meant - but yes chaining over multiple lines needs the `.` at the end of the line. TS for example it doesn't matter or can be in the beginning of line as well

File: cmd/taskctl/testdata/gha.sample.yml (hunk 3 above)
Comment (ElvenSpellmaker) [reply cmd/taskctl/testdata/gha.sample.yml:84]: @dnitsch

File: docs/v2.md (hunk 1 above)
Comment (ElvenSpellmaker) [reply docs/v2.md:8]: @dnitsch

File: internal/config/context.go [hunk 10]
```diff
@@ -68,15 +68,14 @@ func newEnvFile(defEnvFile *utils.Envfile, isContainerContext bool) (*utils.Envf
 	}
 
 	envFile := utils.NewEnvFile(func(e *utils.Envfile) {
-		e.Generate = defEnvFile.Generate
-		if isContainerContext {
-			e.Generate = true
-		}
+		// REMOVE Generate - as we will always generate when the context is container
```
Comment (ElvenSpellmaker) [inline internal/config/context.go:71]: Did you remove it?

File: internal/genci/githubimpl.go (hunk 5 above)
Comment (ElvenSpellmaker) [reply internal/genci/githubimpl.go:169]: @dnitsch That's a terrible lint pattern XD

File: pkg/runner/context.go [hunk 11]
```diff
@@ -169,7 +167,15 @@ func (c *ExecutionContext) GenerateEnvfile(env *variables.Variables) error {
 	builder := []string{}
 	// iterate through all of the environment variables and add the selected ones to the builder
 	// env container at this point should already include all the merged variables by precedence
+
 	// TODO: if envfile path was provided we should merge it in with Env and inject as a whole into the container
```
Comment (ElvenSpellmaker) [inline pkg/runner/context.go:171]: Is this being resolved in this?

File: pkg/runner/runner.go [hunk 12]
```diff
@@ -164,7 +179,7 @@ func (r *TaskRunner) Run(t *task.Task) error {
 	if err != nil {
 		return err
 	}
-
+	//
```
Comment (ElvenSpellmaker) [inline pkg/runner/runner.go:182]: Nice comment

File: pkg/runner/runner_test.go [hunk 13]
```diff
@@ -104,52 +106,127 @@ func TestTaskRunner(t *testing.T) {
...
+
+		tf, err := os.CreateTemp("", "exclude-*.env")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// for _, v := range [][2]string{{"ADDED", "/old/foo"}, {"NEW_STUFF", "/old/bar"}} {
+		// 	os.Setenv(v[0], v[1])
+		// }
+		// defer os.Clearenv()
```
Comment (ElvenSpellmaker) [inline pkg/runner/runner_test.go:174]: Needed?

File: pkg/runner/runner_test.go [hunk 14]
```diff
@@ -272,3 +349,30 @@ func TestRunner_Run_with_Artifacts(t *testing.T) {
...
+	task.EnvFile = utils.NewEnvFile().WithPath(tf.Name())
+	task.Commands = []string{"true"}
+
+	err = tr.Run(task)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// if got := task.Env.Get("FOO"); got != "bar" {
+	// 	t.Errorf("FOO env var not merged from file, got %s, wanted bar", got)
+	// }
```
Comment (ElvenSpellmaker) [inline pkg/runner/runner_test.go:377]: Needed?

File: tasks.yaml (hunk 9 above)
Comment (ElvenSpellmaker) [reply tasks.yaml:201]: @dnitsch

**Review by ElvenSpellmaker — COMMENTED** [review]
Mostly OK, please address things here

File: internal/config/context.go (hunk 10 above)
Comment (dnitsch) [reply internal/config/context.go:71]: yeah it's been removed - I suppose the comment should read `REMOVED` "¯\_(ツ)_/¯"

**Review by ElvenSpellmaker — APPROVED** [review]

**Pull request merged**